opengever/maintenance/browser/warmup.py

```diff
@@ -30,54 +44,170 @@ class WarmupView(BrowserView):
-            repo = brain.getObject()
+            brain.getObject()
 
     def _warmup_medium(self):
         # Fetch brains for all common types
         for type_iface in COMMON_TYPES:
             count = 0
-            brains = self.catalog(object_provides=type_iface.__identifier__)
+            brains = self.catalog.unrestrictedSearchResults(
+                object_provides=type_iface.__identifier__)
             for brain in brains:
                 count += 1
             log.info('Fetched {} brains for type {}'.format(
                 count, type_iface.__identifier__))
 
         self._warmup_minimal()
 
-    def _warmup_full(self):
-        # Fetch brains and objects for all common types
-        for type_iface in COMMON_TYPES:
-            count = 0
-            brains = self.catalog(object_provides=type_iface.__identifier__)
-            for brain in brains:
-                obj = brain.getObject()
-                count += 1
-            log.info('Fetched {} brains and objects for type {}'.format(
-                count, type_iface.__identifier__))
+    def _warmup_catalog(self, zctext_indexes=True, lexicons=True):
+        start = time.time()
 
-        self._warmup_medium()
-        self._warmup_minimal()
+        self._log_cache_stats('before warmup')
+        warmup_catalog(
+            self.catalog, zctext_indexes=zctext_indexes, lexicons=lexicons)
+        self._log_cache_stats('after warmup')
+
+        duration = time.time() - start
+        log.info("Catalog warmup took %.2fs." % duration)
+
+    def _log_cache_stats(self, when):
+        conn = self.context._p_jar
+        db = conn.db()
+        log.info('')
+        log.info("Cache size (# objs) (%s): %s" % (when, db.cacheSize()))
+        log.info("Total estimated cache size (%s): %.2f MiB" %
+                 (when, conn._cache.total_estimated_size / 1024.0 / 1024.0))
+        log.info("Current instance RSS (%s): %.2f MiB" % (
+            when, get_rss() / 1024.0))
+        log.info('')
+
+
+def warmup_catalog(catalog, zctext_indexes=True, lexicons=True):
+    log.info('Loading metadata...')
+    list(catalog._catalog.data.items())
+
+    log.info('Loading indexes...')
+    for index_name in catalog.indexes():
+        index = catalog._catalog.indexes[index_name]
+        if not zctext_indexes and isinstance(index, ZCTextIndex):
+            continue
+        log.info('Loading index %r...' % index_name)
+        warmup_index(index, lexicons=lexicons)
+
+    log.info('Done warming up catalog.')
+
+
+def warmup_index(index, lexicons=True):
+    """Load internal index data structures by iterating over contents of
+    *BTrees and *TreeSets, causing their respective buckets to be loaded.
+    """
+    if isinstance(index, (FieldIndex, DateIndex, KeywordIndex)):
+        # value -> set of docids
+        for key, value in index._index.items():
+            list(value)
+
+        # docid -> field value
+        list(index._unindex.items())
```

Comment:
One could skip warming up the reverse index (`_unindex`) as it shouldn't be used for querying the catalog. But indexing an object requires it. I'm not sure about the performance impact of a cold reverse index when indexing an object.